File: c_programs/Data Structure Programs/Matrix/transmat.c
/*WAP in C to find the transpose of a matrix*/

#include<stdio.h>
#include<stdlib.h>
//function to take input in a matrix
void getmatrix(int (*arr)[20],int row,int column)
{
	int i,j;
	for(i=0;i<row;i++)
	{
		for(j=0;j<column;j++)
			scanf("%d",&arr[i][j]);
	}
}
//function to display a matrix
void dispmatrix(int (*arr)[20],int row,int column)
{
	int i,j;
	for(i=0;i<row;i++)
	{
		for(j=0;j<column;j++)
			printf("%d\t",arr[i][j]);
		printf("\n");
	}
}
//function for interchanging variable values
void swap(int*a,int*b)
{
	*a=*a+*b;
	*b=*a-*b;
	*a=*a-*b;
}
//function for input validation
void validate(int row,int col)
{
    if(row<1 || col<1)
    {
        printf("Invalid Dimensions\nPlease Enter a Valid Dimension");
        exit(1);
    }
}


void trans(int(*arr)[20],int row,int col)
{
	int i,j,mid;
	mid=0;
	while(mid<row) //traversing the square part of the input matrix
	{
		for(i=mid+1;i<row;i++) 
			swap(&arr[mid][i],&arr[i][mid]);
		mid++;
	}
	if(row!=col) //if input matrix is not a square matrix
	{
		j=row;
		while(j<col)//traversing the non square part of the matrix
		{
			for(i=0;i<row;i++)
				swap(&arr[i][j],&arr[j][i]);
			j++;
		}
	}
}

int main(void)
{
	int arr[20][20],row,col;
	printf("Enter the number of rows needed: ");
	scanf("%d",&row);
	printf("Enter the number of columns needed: ");
	scanf("%d",&col);
    validate(row,col); //input validation
	printf("Enter %d elements of the matrix: \n",row*col);
	getmatrix(arr,row,col);
	printf("\nElements of the matrix: \n");
	dispmatrix(arr,row,col); //displaying input matrix
	trans(arr,row,col);
    printf("\nTranspose: \n");
	dispmatrix(arr,col,row);//displaying transposed input matrix
	return 0;
}

File: python/assigns/ds/transmat.py
def getmatrix(row,col):
    if col is None:
        col=row
    mat=[]
    for i in range(row):
        arr=[]
        for j in range(col):
            temp=int(input(f"Enter the {i},{j} element: "))
            arr.append(temp)
        mat.append(arr)
    return mat

def showmatrix(mat):
    row=len(mat)
    col=len(mat[0])
    for i in range(row):
        for j in range(col):
            print(mat[i][j],end="\t")
        print()

def pad(mat,row,col):
    diff=abs(row-col)
    if row < col:
        for i in range(diff):
            mat.append([0]*col)
    else:
        for val in mat:
            for i in range(diff):
                val.append(0)
    return mat

def transposemat(mat):
    row=len(mat)
    col=len(mat[0])
    if row != col:
        mat=pad(mat,row,col)
    for i in range(row):
        for j in range(col):
            if j>i:
                mat[i][j],mat[j][i]=mat[j][i],mat[i][j]
    return mat
    
row,col=int(input("Enter number of rows: ")),int(input("Enter number of columns: "))
print(f"Input Dimensions= {row}x{col}")

mat=getmatrix(row,col)
print("State of the matrix: ")
showmatrix(mat)
mat=transposemat(mat)
print("Transpose:")
showmatrix(mat)
